File: api-openbis-java/source/java/ch/ethz/sis/openbis/generic/imagingapi/v3/dto/ImagingDataSetImage.java
/*
 *  Copyright ETH 2023 Zürich, Scientific IT Services
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package ch.ethz.sis.openbis.generic.imagingapi.v3.dto;

import ch.ethz.sis.openbis.generic.asapi.v3.dto.common.property.PropertiesDeserializer;
import ch.systemsx.cisd.base.annotation.JsonObject;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import java.io.Serializable;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@JsonObject("imaging.dto.ImagingDataSetImage")
public class ImagingDataSetImage implements Serializable
{
    private static final long serialVersionUID = 1L;

    @JsonProperty
    private ImagingDataSetConfig config;

    @JsonProperty
    private List<ImagingDataSetPreview> previews;

    @JsonProperty
    @JsonDeserialize(contentUsing = PropertiesDeserializer.class)
    private Map<String, Serializable> imageConfig;

    @JsonProperty
    private Integer index;

    @JsonProperty
    @JsonDeserialize(contentUsing = PropertiesDeserializer.class)
    private Map<String, Serializable> metadata;

    @JsonIgnore
    public List<ImagingDataSetPreview> getPreviews()
    {
        return previews;
    }

    public void setPreviews(
            List<ImagingDataSetPreview> previews)
    {
        this.previews = previews;
    }

    @JsonIgnore
    public Map<String, Serializable> getImageConfig()
    {
        if(imageConfig == null)
        {
            imageConfig = new HashMap<>();
        }
        return imageConfig;
    }

    public void setImageConfig(Map<String, Serializable> imageConfig)
    {
        this.imageConfig = imageConfig;
    }

    @JsonIgnore
    public Integer getIndex()
    {
        return index;
    }

    public void setIndex(Integer index)
    {
        this.index = index;
    }

    @JsonIgnore
    public Map<String, Serializable> getMetadata()
    {
        if(metadata == null)
        {
            metadata = new HashMap<>();
        }
        return metadata;
    }

    public void setMetadata(Map<String, Serializable> metadata)
    {
        this.metadata = metadata;
    }

    @JsonIgnore
    public ImagingDataSetConfig getConfig()
    {
        return config;
    }

    public void setConfig(ImagingDataSetConfig config)
    {
        this.config = config;
    }

    @Override
    public String toString()
    {
        return "ImagingDataSetImage:";
    }

}

File: api-openbis-javascript/src/v3/imaging/dto/ImagingDataSetImage.js
define([ "stjs", "util/Exceptions" ], function(stjs, exceptions) {
	var ImagingDataSetImage = function() {
	};
	stjs.extend(ImagingDataSetImage, null, [], function(constructor, prototype) {
		prototype['@type'] = 'imaging.dto.ImagingDataSetImage';
		constructor.serialVersionUID = 1;
		prototype.config = null;
		prototype.previews = null;
		prototype.imageConfig = null;
		prototype.index = null;
		prototype.metadata = null;

		prototype.getPreviews = function() {
			return this.previews;
		};
		prototype.setPreviews = function(previews) {
			this.previews = previews;
		};
		prototype.getConfig = function() {
            return this.config;
        };
        prototype.setConfig = function(config) {
            this.config = config;
        };
        prototype.getIndex = function() {
            return this.index;
        };
        prototype.setIndex = function(index) {
            this.index = index;
        };
		prototype.getMetadata = function() {
			return this.metadata;
		};
		prototype.setMetadata = function(metadata) {
			this.metadata = metadata;
		};
		prototype.getImageConfig = function() {
            return this.imageConfig;
        };
        prototype.setImageConfig = function(imageConfig) {
            this.imageConfig = imageConfig;
        };
		prototype.toString = function() {
            return "ImagingDataSetImage: " + this.previews;
        };

	}, {
		previews : {
            name : "List",
            arguments : [ "ImagingDataSetPreview"]
        },
        imageConfig : {
            name : "Map",
            arguments : [ "String", "Serializable" ]
        },
        metadata : {
            name : "Map",
            arguments : [ "String", "Serializable" ]
        }
	});
	return ImagingDataSetImage;
})